src/main/java/seedu/address/model/ModelManager.java

```diff
@@ -79,6 +88,49 @@ public synchronized void addTask(Task person) {
         indicateTaskBookChanged();
     }
 
+    //=============for undo and redo===================================
+    @Override
+    public Command undo() throws IllegalValueException {
+
+        if (head <= 0) {
+            System.out.println(head);
+            throw new IllegalValueException("No actions to undo");
+        }
+        head--;
+        Commit commit = commits.get(head);
+        resetData(new TaskBook(commit.getTaskBook()));
+        return commit.getCommand();
+    }
+
+    @Override
+    public void recordState(Command command) {
+        commits.add(new Commit(command, new TaskBook(getTaskBook())));
+        this.head = commits.size() - 1;
```

Comment:
> undo is a command,

I don't understand this argument. To me, `ModelManager::undo()` looks like a method to me that can be called by any code, not only commands. The bare minimum API design needs to take that into account.

Also, `ModelManager::recordState()` is also a method that can be called by any code, not only commands.